### PR TITLE
Adds cuckoo optimiser

### LINKED ISSUE #307: Add cuckoo optimiser
### Feature description

Cuckoo search is a gradient-free optimisation method capable of global optimisation guarentees [1]. Various implementations have shown improvements over CMA-ES and PSO [2]. This issue is create a PyBOP implementation of CSA.

[1] https://www.sciencedirect.com/science/article/pii/S096007791100107X
[2] https://www.sciencedirect.com/science/article/pii/S0957417422007679

### Motivation

_No response_

### Possible implementation

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by BradyPlanden**
# Description

This PR adds a cuckoo search implementation and corresponding tests.

## Issue reference
Fixes #307 

## Review
Before you mark your PR as ready for review, please ensure that you've considered the following:
- Updated the [CHANGELOG.md](https://github.com/pybop-team/PyBOP/blob/develop/CHANGELOG.md) in reverse chronological order (newest at the top) with a concise description of the changes, including the PR number.
- Noted any breaking changes, including details on how it might impact existing functionality.

## Type of change
- [X] New Feature: A non-breaking change that adds new functionality.
- [ ] Optimization: A code change that improves performance.
- [ ] Examples: A change to existing or additional examples.
- [X] Bug Fix: A non-breaking change that addresses an issue.
- [ ] Documentation: Updates to documentation or new documentation for new features.
- [ ] Refactoring: Non-functional changes that improve the codebase.
- [ ] Style: Non-functional changes related to code style (formatting, naming, etc).
- [X] Testing: Additional tests to improve coverage or confirm functionality.
- [ ] Other: (Insert description of change)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybop-team/PyBOP/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All unit tests pass: `$ nox -s tests`
- [ ] The documentation builds: `$ nox -s doctest`

You can run integration tests, unit tests, and doctests together at once, using `$ nox -s quick`.

## Further checks:
- [ ] Code is well-commented, especially in complex or unclear areas.
- [ ] Added tests that prove my fix is effective or that my feature works.
- [ ] Checked that coverage remains or improves, and added tests if necessary to maintain or increase coverage.

Thank you for contributing to our project! Your efforts help us to deliver great software.

### CODE REVIEW

File: pybop/optimisers/_cuckoo.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+        u = random.normal(0, sigma_u, size=size)
+        v = random.normal(0, sigma_v, size=size)
+        step = u / power(abs(v), 1 / alpha)
+
+        return step
+
+    def abandon_nests(self, idx):
+        """
+        Set the boundaries for the parameter space.
```
Comment (BradyPlanden) [inline pybop/optimisers/_cuckoo.py:147]:
```suggestion
        Updates the nests to abandon the worst performers and reinitialise. 
```

File: pybop/optimisers/_cuckoo.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+            self._nests[idx] = np.random.uniform(
+                low=self._boundaries.lower(),
+                high=self._boundaries.upper(),
+            )
+        else:
+            self._nests[idx] = np.random.normal(self._x0, self._sigma0)
+
+    def clip_nests(self, x):
+        """
+        Clip the input array to the boundaries.
```
Comment (BradyPlanden) [inline pybop/optimisers/_cuckoo.py:159]:
```suggestion
        Clip the input array to the boundaries if available.
```

File: pybop/optimisers/pints_optimisers.py [hunk 3]
```diff
@@ -240,3 +240,31 @@ def __init__(self, cost, **optimiser_kwargs):
                 + "Please choose another optimiser."
             )
         super().__init__(cost, PintsCMAES, **optimiser_kwargs)
+
+
+class CuckooSearch(BasePintsOptimiser):
+    """
+    Adapter for the Cuckoo Search optimiser in PyBOP.
+
+    Cuckoo Search is a population-based optimization algorithm inspired by the brood parasitism of some cuckoo species.
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:249]:
```suggestion
    Cuckoo Search is a population-based optimisation algorithm inspired by the brood parasitism of some cuckoo species.
```

File: pybop/optimisers/pints_optimisers.py [hunk 4]
```diff
@@ -240,3 +240,31 @@ def __init__(self, cost, **optimiser_kwargs):
...
+    It is designed to be simple, efficient, and robust, and is suitable for global optimization problems.
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:250]:
```suggestion
    It is designed to be simple, efficient, and robust, and is suitable for global optimisation problems.
```

File: pybop/optimisers/pints_optimisers.py [hunk 5]
```diff
@@ -240,3 +240,31 @@ def __init__(self, cost, **optimiser_kwargs):
...
+
+    Parameters
+    ----------
+    **optimiser_kwargs : optional
+        Valid PyBOP option keys and their values, for example:
+        x0 : array_like
+            Initial
```
Comment (BradyPlanden) [inline pybop/optimisers/pints_optimisers.py:257]:
```suggestion
            Initial parameter values.
```

File: pybop/optimisers/_cuckoo.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+            )
+
+        self._fitness = np.full(self._n, np.inf)
+
+        # Initialise best solutions
+        self._x_best = np.copy(x0)
+        self._f_best = np.inf
+
+        # Set iteration count
+        self._iterations = 1
```
Comment (NicolaCourtier) [inline pybop/optimisers/_cuckoo.py:85]: I'm not sure if this should count as the first iteration, with an infinite cost?
I see that `self._iterations` is used in line 97 below, but maybe set it equal to zero here and add 1 to line 97 for consistency.

**Review by NicolaCourtier — APPROVED** [review]
Looks good to me, don't forget to update the Changelog!

**Pull request merged**